File: binding/src/libqregion.cpp
#include <QBitmap>
#include <QPoint>
#include <QRect>
#include <QRegion>
#include <QVariant>
#include <qregion.h>
#include "libqregion.h"
#include "libqregion.hxx"

QRegion* QRegion_new() {
	 return new QRegion();
}

QRegion* QRegion_new2(int x, int y, int w, int h) {
	 return new QRegion(x, y, w, h);
}

QRegion* QRegion_new3(const QRect* r) {
	 return new QRegion(*r);
}

QRegion* QRegion_new4(const QRegion* region) {
	 return new QRegion(*region);
}

QRegion* QRegion_new5(const QBitmap* bitmap) {
	 return new QRegion(*bitmap);
}

QRegion* QRegion_new6(int x, int y, int w, int h, int t) {
	 return new QRegion(x, y, w, h, static_cast<QRegion::RegionType>(t));
}

QRegion* QRegion_new7(const QRect* r, int t) {
	 return new QRegion(*r, static_cast<QRegion::RegionType>(t));
}

void QRegion_OperatorAssign(QRegion* self, const QRegion* param1) {
	self->operator=(*param1);
}

void QRegion_Swap(QRegion* self, QRegion* other) {
	self->swap(*other);
}

bool QRegion_IsEmpty(const QRegion* self) {
	return self->isEmpty();
}

bool QRegion_IsNull(const QRegion* self) {
	return self->isNull();
}

const QRect** QRegion_Begin(const QRegion* self) {
	return self->begin();
}

const QRect** QRegion_Cbegin(const QRegion* self) {
	return self->cbegin();
}

const QRect** QRegion_End(const QRegion* self) {
	return self->end();
}

const QRect** QRegion_Cend(const QRegion* self) {
	return self->cend();
}

bool QRegion_Contains(const QRegion* self, const QPoint* p) {
	return self->contains(*p);
}

bool QRegion_Contains2(const QRegion* self, const QRect* r) {
	return self->contains(*r);
}

void QRegion_Translate(QRegion* self, int dx, int dy) {
	self->translate(dx, dy);
}

void QRegion_Translate2(QRegion* self, const QPoint* p) {
	self->translate(*p);
}

QRegion* QRegion_Translated(const QRegion* self, int dx, int dy) {
	return new QRegion(self->translated(dx, dy));
}

QRegion* QRegion_Translated2(const QRegion* self, const QPoint* p) {
	return new QRegion(self->translated(*p));
}

QRegion* QRegion_United(const QRegion* self, const QRegion* r) {
	return new QRegion(self->united(*r));
}

QRegion* QRegion_United2(const QRegion* self, const QRect* r) {
	return new QRegion(self->united(*r));
}

QRegion* QRegion_Intersected(const QRegion* self, const QRegion* r) {
	return new QRegion(self->intersected(*r));
}

QRegion* QRegion_Intersected2(const QRegion* self, const QRect* r) {
	return new QRegion(self->intersected(*r));
}

QRegion* QRegion_Subtracted(const QRegion* self, const QRegion* r) {
	return new QRegion(self->subtracted(*r));
}

QRegion* QRegion_Xored(const QRegion* self, const QRegion* r) {
	return new QRegion(self->xored(*r));
}

bool QRegion_Intersects(const QRegion* self, const QRegion* r) {
	return self->intersects(*r);
}

bool QRegion_Intersects2(const QRegion* self, const QRect* r) {
	return self->intersects(*r);
}

QRect* QRegion_BoundingRect(const QRegion* self) {
	return new QRect(self->boundingRect());
}

void QRegion_SetRects(QRegion* self, const QRect* rect, int num) {
	self->setRects(rect, num);
}

void QRegion_SetRects2(QRegion* self, libqt_list r) {
	self->setRects(QSpan<const QRect>());
}

libqt_list QRegion_Rects(const QRegion* self) {
	auto _ret = self->rects();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

int QRegion_RectCount(const QRegion* self) {
	return self->rectCount();
}

QRegion* QRegion_OperatorBitwiseOr(const QRegion* self, const QRegion* r) {
	return new QRegion(self->operator|(*r));
}

QRegion* QRegion_OperatorPlus(const QRegion* self, const QRegion* r) {
	return new QRegion(self->operator+(*r));
}

QRegion* QRegion_OperatorPlus2(const QRegion* self, const QRect* r) {
	return new QRegion(self->operator+(*r));
}

QRegion* QRegion_OperatorBitwiseAnd(const QRegion* self, const QRegion* r) {
	return new QRegion(self->operator&(*r));
}

QRegion* QRegion_OperatorBitwiseAnd2(const QRegion* self, const QRect* r) {
	return new QRegion(self->operator&(*r));
}

QRegion* QRegion_OperatorMinus(const QRegion* self, const QRegion* r) {
	return new QRegion(self->operator-(*r));
}

QRegion* QRegion_OperatorBitwiseNot(const QRegion* self, const QRegion* r) {
	return new QRegion(self->operator^(*r));
}

void QRegion_OperatorBitwiseOrAssign(QRegion* self, const QRegion* r) {
	self->operator|=(*r);
}

QRegion* QRegion_OperatorPlusAssign(QRegion* self, const QRegion* r) {
	return new QRegion(self->operator+=(*r));
}

QRegion* QRegion_OperatorPlusAssign2(QRegion* self, const QRect* r) {
	return new QRegion(self->operator+=(*r));
}

void QRegion_OperatorBitwiseAndAssign(QRegion* self, const QRegion* r) {
	self->operator&=(*r);
}

void QRegion_OperatorBitwiseAndAssign2(QRegion* self, const QRect* r) {
	self->operator&=(*r);
}

QRegion* QRegion_OperatorMinusAssign(QRegion* self, const QRegion* r) {
	return new QRegion(self->operator-=(*r));
}

void QRegion_OperatorBitwiseXorAssign(QRegion* self, const QRegion* r) {
	self->operator^=(*r);
}

bool QRegion_OperatorEqual(const QRegion* self, const QRegion* r) {
	return self->operator==(*r);
}

bool QRegion_OperatorNotEqual(const QRegion* self, const QRegion* r) {
	return self->operator!=(*r);
}

QVariant* QRegion_OperatorQVariant(const QRegion* self) {
	return new QVariant(self->operator QVariant());
}

struct HRGN__* QRegion_ToHRGN(const QRegion* self) {
	return self->toHRGN();
}

QRegion* QRegion_FromHRGN(struct HRGN__* hrgn) {
	return new QRegion(QRegion::fromHRGN(hrgn));
}

void QRegion_Delete(QRegion* self) {
    delete self;
}

File: binding/include/libqregion.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qregion.html
pub const qregion = struct {

    /// New constructs a new QRegion object.
    pub fn New() QtC.QRegion {
        return qtc.QRegion_new();
    }


    /// New2 constructs a new QRegion object.
    pub fn New2(x: i32, y: i32, w: i32, h: i32) QtC.QRegion {
        return qtc.QRegion_new2(x, y, w, h);
    }


    /// New3 constructs a new QRegion object.
    pub fn New3(r: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_new3(@ptrCast(r));
    }


    /// New4 constructs a new QRegion object.
    pub fn New4(pa: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_new4(@ptrCast(pa));
    }


    /// New5 constructs a new QRegion object.
    pub fn New5(region: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_new5(@ptrCast(region));
    }


    /// New6 constructs a new QRegion object.
    pub fn New6(bitmap: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_new6(@ptrCast(bitmap));
    }


    /// New7 constructs a new QRegion object.
    pub fn New7(x: i32, y: i32, w: i32, h: i32, t: i32) QtC.QRegion {
        return qtc.QRegion_new7(x, y, w, h, @intCast(t));
    }


    /// New8 constructs a new QRegion object.
    pub fn New8(r: ?*anyopaque, t: i32) QtC.QRegion {
        return qtc.QRegion_new8(@ptrCast(r), @intCast(t));
    }


    /// New9 constructs a new QRegion object.
    pub fn New9(pa: ?*anyopaque, fillRule: i32) QtC.QRegion {
        return qtc.QRegion_new9(@ptrCast(pa), @intCast(fillRule));
    }


    pub fn OperatorAssign(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QRegion_OperatorAssign(@ptrCast(self), @ptrCast(param1));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QRegion_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn IsEmpty(self: ?*anyopaque, ) bool {
        return qtc.QRegion_IsEmpty(@ptrCast(self));
    }

    pub fn IsNull(self: ?*anyopaque, ) bool {
        return qtc.QRegion_IsNull(@ptrCast(self));
    }

    pub fn Begin(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QRegion_Begin(@ptrCast(self));
    }

    pub fn Cbegin(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QRegion_Cbegin(@ptrCast(self));
    }

    pub fn End(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QRegion_End(@ptrCast(self));
    }

    pub fn Cend(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QRegion_Cend(@ptrCast(self));
    }

    pub fn Rbegin(self: ?*anyopaque, ) std::reverse_iterator<const QRect *> {
        return qtc.QRegion_Rbegin(@ptrCast(self));
    }

    pub fn Crbegin(self: ?*anyopaque, ) std::reverse_iterator<const QRect *> {
        return qtc.QRegion_Crbegin(@ptrCast(self));
    }

    pub fn Rend(self: ?*anyopaque, ) std::reverse_iterator<const QRect *> {
        return qtc.QRegion_Rend(@ptrCast(self));
    }

    pub fn Crend(self: ?*anyopaque, ) std::reverse_iterator<const QRect *> {
        return qtc.QRegion_Crend(@ptrCast(self));
    }

    pub fn Contains(self: ?*anyopaque, p: ?*anyopaque) bool {
        return qtc.QRegion_Contains(@ptrCast(self), @ptrCast(p));
    }

    pub fn Contains2(self: ?*anyopaque, r: ?*anyopaque) bool {
        return qtc.QRegion_Contains2(@ptrCast(self), @ptrCast(r));
    }

    pub fn Translate(self: ?*anyopaque, dx: i32, dy: i32) void {
        qtc.QRegion_Translate(@ptrCast(self), dx, dy);
    }

    pub fn Translate2(self: ?*anyopaque, p: ?*anyopaque) void {
        qtc.QRegion_Translate2(@ptrCast(self), @ptrCast(p));
    }

    pub fn Translated(self: ?*anyopaque, dx: i32, dy: i32) QtC.QRegion {
        return qtc.QRegion_Translated(@ptrCast(self), dx, dy);
    }

    pub fn Translated2(self: ?*anyopaque, p: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_Translated2(@ptrCast(self), @ptrCast(p));
    }

    pub fn United(self: ?*anyopaque, r: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_United(@ptrCast(self), @ptrCast(r));
    }

    pub fn United2(self: ?*anyopaque, r: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_United2(@ptrCast(self), @ptrCast(r));
    }

    pub fn Intersected(self: ?*anyopaque, r: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_Intersected(@ptrCast(self), @ptrCast(r));
    }

    pub fn Intersected2(self: ?*anyopaque, r: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_Intersected2(@ptrCast(self), @ptrCast(r));
    }

    pub fn Subtracted(self: ?*anyopaque, r: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_Subtracted(@ptrCast(self), @ptrCast(r));
    }

    pub fn Xored(self: ?*anyopaque, r: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_Xored(@ptrCast(self), @ptrCast(r));
    }

    pub fn Intersects(self: ?*anyopaque, r: ?*anyopaque) bool {
        return qtc.QRegion_Intersects(@ptrCast(self), @ptrCast(r));
    }

    pub fn Intersects2(self: ?*anyopaque, r: ?*anyopaque) bool {
        return qtc.QRegion_Intersects2(@ptrCast(self), @ptrCast(r));
    }

    pub fn BoundingRect(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QRegion_BoundingRect(@ptrCast(self));
    }

    pub fn SetRects(self: ?*anyopaque, rect: ?*anyopaque, num: i32) void {
        qtc.QRegion_SetRects(@ptrCast(self), @ptrCast(rect), num);
    }

    pub fn SetRects2(self: ?*anyopaque, r: QSpan<const QRect>) void {
        qtc.QRegion_SetRects2(@ptrCast(self), r);
    }

    pub fn Rects(self: ?*anyopaque, ) QSpan<const QRect> {
        return qtc.QRegion_Rects(@ptrCast(self));
    }

    pub fn RectCount(self: ?*anyopaque, ) i32 {
        return qtc.QRegion_RectCount(@ptrCast(self));
    }

    pub fn OperatorBitwiseOr(self: ?*anyopaque, r: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_OperatorBitwiseOr(@ptrCast(self), @ptrCast(r));
    }

    pub fn OperatorPlus(self: ?*anyopaque, r: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_OperatorPlus(@ptrCast(self), @ptrCast(r));
    }

    pub fn OperatorPlus2(self: ?*anyopaque, r: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_OperatorPlus2(@ptrCast(self), @ptrCast(r));
    }

    pub fn OperatorBitwiseAnd(self: ?*anyopaque, r: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_OperatorBitwiseAnd(@ptrCast(self), @ptrCast(r));
    }

    pub fn OperatorBitwiseAnd2(self: ?*anyopaque, r: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_OperatorBitwiseAnd2(@ptrCast(self), @ptrCast(r));
    }

    pub fn OperatorMinus(self: ?*anyopaque, r: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_OperatorMinus(@ptrCast(self), @ptrCast(r));
    }

    pub fn OperatorBitwiseNot(self: ?*anyopaque, r: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_OperatorBitwiseNot(@ptrCast(self), @ptrCast(r));
    }

    pub fn OperatorBitwiseOrAssign(self: ?*anyopaque, r: ?*anyopaque) void {
        qtc.QRegion_OperatorBitwiseOrAssign(@ptrCast(self), @ptrCast(r));
    }

    pub fn OperatorPlusAssign(self: ?*anyopaque, r: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_OperatorPlusAssign(@ptrCast(self), @ptrCast(r));
    }

    pub fn OperatorPlusAssign2(self: ?*anyopaque, r: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_OperatorPlusAssign2(@ptrCast(self), @ptrCast(r));
    }

    pub fn OperatorBitwiseAndAssign(self: ?*anyopaque, r: ?*anyopaque) void {
        qtc.QRegion_OperatorBitwiseAndAssign(@ptrCast(self), @ptrCast(r));
    }

    pub fn OperatorBitwiseAndAssign2(self: ?*anyopaque, r: ?*anyopaque) void {
        qtc.QRegion_OperatorBitwiseAndAssign2(@ptrCast(self), @ptrCast(r));
    }

    pub fn OperatorMinusAssign(self: ?*anyopaque, r: ?*anyopaque) QtC.QRegion {
        return qtc.QRegion_OperatorMinusAssign(@ptrCast(self), @ptrCast(r));
    }

    pub fn OperatorBitwiseXorAssign(self: ?*anyopaque, r: ?*anyopaque) void {
        qtc.QRegion_OperatorBitwiseXorAssign(@ptrCast(self), @ptrCast(r));
    }

    pub fn OperatorEqual(self: ?*anyopaque, r: ?*anyopaque) bool {
        return qtc.QRegion_OperatorEqual(@ptrCast(self), @ptrCast(r));
    }

    pub fn OperatorNotEqual(self: ?*anyopaque, r: ?*anyopaque) bool {
        return qtc.QRegion_OperatorNotEqual(@ptrCast(self), @ptrCast(r));
    }

    pub fn OperatorQVariant(self: ?*anyopaque, ) QtC.QVariant {
        return qtc.QRegion_OperatorQVariant(@ptrCast(self));
    }

    pub fn ToHRGN(self: ?*anyopaque, ) struct HRGN__ {
        return @ptrCast(qtc.QRegion_ToHRGN(@ptrCast(self)));
    }

    pub fn FromHRGN(hrgn: *struct HRGN__) QtC.QRegion {
        return qtc.QRegion_FromHRGN(@ptrCast(hrgn));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QRegion_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/region.html#types
pub const enums = struct {
    pub const RegionType = enum {
        pub const Rectangle: i32 = 0;
        pub const Ellipse: i32 = 1;
    };

};
